doc/developer/design/aphyr-txn-notes.md

```diff
@@ -0,0 +1,115 @@
+the sequencer) shares the order of execution it determined with other all other
+the services that depend on that order (known as scheulders). Because the
+execution is deterministic, ensuring each scheduler receives the same set of
+inputs means they will eventually converge to the same final states.
+
+## Materialize design
+
+Materialize propagates all write operations to a durable WAL. (Note that unlike
+Calvin, we only propagate writes to the WAL and not the entire transactional
+input; we are not running full stack of Materialize services vertically.)
+
+Storage nodes are then responsible for taking updates to the WAL and attempting
+to apply them to their local state--note that this, in essence, pushes down
+transaction logic to the storage layer. Storage nodes then need some means to
+communicate with one another to either all accept or all reject the update. In
+the case of rejection, the storage node needs some means to discard the
+provisional writes.
+
+Note that in this design, some external source is responsible for determining
+the order in which writes get appended to the WAL, i.e. the WAL itself is truly
+just a log that offers to real invariants itself.
+
+This design is inspired by Calvin in as much as the storage nodes get to behave
+akin t
+
+### Example architecture
+
+```
+  client
+  |    ^
+ txnA  |
+  |   res
+  v    |                 storageA    storageB
+[ADAPTER]-wA->[WAL]<-get---?
+^              ^-----------get-----------?
+|
+|              if a?.is_ok() && b?.is_ok() {
+|                  return OK to client
+|              } else {
+|                  rollback(A);
+|                  rollback(B);
+|                  return ABORT to client
+|______________}
+```
+
+### Questions
+
+- How do writes invalidate outstanding reads in e.g. `UPDATE`? I realize this
+  design says that sequencing has to happen elsewhere, but just curious where we
+  envision that potentially being.
+- How do e.g. column constraints get handled by the txn layer? Seems like this
+  moves "business logic" into storage, which I don't think will make the storage
+  people happy.
+- How do we rollback if we're already in the storage layer?
+
+# CAS
+
+## Background
+
+This is the design Kyle has been moving forward with fleshing out, which relies
+on a the WAL having a timestamp only ever written to atomically.
+
+## Materialize design
+
+Each writable TVC will have a WAL that acts as a buffer for writes before being
+committed to storage. Each WAL will have a timestamp indicating the last time at
+which an operation occurred; we'l call this the WAL ts for simplicity.
+
+When users want to write to a TVC, they'll use `append(updates, from ts, new
+frontier)`. This implementation of `append` will succeed iff it's able to update
+the WAL's ts using an atomic CAS, i.e. the WAL's current ts is `from ts`.
+
+When reading, the ADAPTER consults the WAL and gets its current timestamp
```

Comment:
I do not recall discussing this, that reads ever need to consult the WAL for a timestamp.